File: packages/exochain-py/exochain/governance/decision.py
"""Governance decisions — titled proposals with collected votes and quorum checks."""

from __future__ import annotations

from enum import StrEnum
from hashlib import sha256

from pydantic import BaseModel, ConfigDict, Field

from ..errors import GovernanceError
from ..types import Did, QuorumResult
from .vote import Vote, VoteChoice


class DecisionStatus(StrEnum):
    """Lifecycle status for a :class:`Decision`."""

    PROPOSED = "proposed"
    DELIBERATING = "deliberating"
    APPROVED = "approved"
    REJECTED = "rejected"
    CHALLENGED = "challenged"


class Decision(BaseModel):
    """A governance decision with accumulated votes.

    Decisions are mutable only via :meth:`cast_vote` (which rejects duplicate
    voters) and by setting :attr:`status`. The :attr:`decision_id` is a
    deterministic 16-hex-char prefix of SHA-256 over title/description/proposer.
    """

    model_config = ConfigDict(validate_assignment=True)

    decision_id: str
    title: str
    description: str
    proposer: Did
    status: DecisionStatus = DecisionStatus.PROPOSED
    decision_class: str | None = None
    votes: list[Vote] = Field(default_factory=list)

    def cast_vote(self, vote: Vote) -> None:
        """Append ``vote`` to this decision. Raises if the voter already voted.

        Raises:
            GovernanceError: if the voter has already cast a vote.
        """
        if any(v.voter == vote.voter for v in self.votes):
            raise GovernanceError(f"voter {vote.voter} has already cast a vote")
        self.votes.append(vote)

    def check_quorum(self, threshold: int) -> QuorumResult:
        """Tally votes and report whether approvals meet ``threshold``."""
        if not isinstance(threshold, int) or isinstance(threshold, bool) or threshold < 0:
            raise GovernanceError("threshold must be a non-negative integer")

        approvals = sum(1 for v in self.votes if v.choice == VoteChoice.APPROVE)
        rejections = sum(1 for v in self.votes if v.choice == VoteChoice.REJECT)
        abstentions = sum(1 for v in self.votes if v.choice == VoteChoice.ABSTAIN)
        total_votes = len(self.votes)

        return QuorumResult(
            met=approvals >= threshold,
            threshold=threshold,
            total_votes=total_votes,
            approvals=approvals,
            rejections=rejections,
            abstentions=abstentions,
        )


class DecisionBuilder:
    """Fluent builder for a :class:`Decision`."""

    def __init__(self, title: str, description: str, proposer: Did) -> None:
        self._title = title
        self._description = description
        self._proposer = proposer
        self._decision_class: str | None = None

    def decision_class(self, cls: str) -> DecisionBuilder:
        """Attach an optional free-form classification label."""
        if not isinstance(cls, str) or not cls.strip():
            raise GovernanceError("decision_class must be a non-empty string")
        self._decision_class = cls
        return self

    def build(self) -> Decision:
        """Validate and produce a :class:`Decision`."""
        if not isinstance(self._title, str) or not self._title.strip():
            raise GovernanceError("title must be non-empty")
        if not isinstance(self._description, str):
            raise GovernanceError("description must be a string")

        payload = f"{self._title}\x00{self._description}\x00{self._proposer}"
        digest = sha256(payload.encode("utf-8")).hexdigest()
        decision_id = digest[:16]

        return Decision(
            decision_id=decision_id,
            title=self._title,
            description=self._description,
            proposer=self._proposer,
            decision_class=self._decision_class,
        )


__all__ = ["Decision", "DecisionBuilder", "DecisionStatus"]

File: packages/exochain-sdk/src/governance/decision.ts
/**
 * Governance decisions — build, cast votes, check quorum.
 *
 * Mirrors the Rust SDK's `governance` module. Decision IDs are content-
 * addressed (SHA-256 over the canonical title/description/proposer payload),
 * votes are appended in-order, and duplicate voters are rejected.
 */

import { GovernanceError } from '../errors.js';
import type { Did, Hash256, QuorumResult } from '../types.js';
import { validateDid } from '../identity/did.js';
import { sha256, bytesToHex } from '../crypto/hash.js';
import { Vote } from './vote.js';

/** Lifecycle states a decision may be in. */
export type DecisionStatus =
  | 'proposed'
  | 'deliberating'
  | 'approved'
  | 'rejected'
  | 'challenged';

/** A full governance decision with accumulated votes. */
export class Decision {
  public readonly decisionId: Hash256;
  public readonly title: string;
  public readonly description: string;
  public readonly proposer: Did;
  public status: DecisionStatus;
  public readonly class?: string;
  readonly #votes: Vote[];

  constructor(args: {
    decisionId: Hash256;
    title: string;
    description: string;
    proposer: Did;
    status?: DecisionStatus;
    class?: string;
  }) {
    this.decisionId = args.decisionId;
    this.title = args.title;
    this.description = args.description;
    this.proposer = args.proposer;
    this.status = args.status ?? 'proposed';
    if (args.class !== undefined) {
      this.class = args.class;
    }
    this.#votes = [];
  }

  /** Read-only snapshot of votes cast so far. */
  public get votes(): readonly Vote[] {
    return this.#votes;
  }

  /**
   * Append a vote. Throws {@link GovernanceError} if the voter has already
   * voted on this decision.
   */
  public castVote(vote: Vote): void {
    for (const existing of this.#votes) {
      if (existing.voter === vote.voter) {
        throw new GovernanceError(`voter ${vote.voter} has already cast a vote`);
      }
    }
    this.#votes.push(vote);
  }

  /**
   * Tally the votes and report whether the approval count meets `threshold`.
   */
  public checkQuorum(threshold: number): QuorumResult {
    if (!Number.isInteger(threshold) || threshold < 0) {
      throw new GovernanceError('threshold must be a non-negative integer');
    }
    let approvals = 0;
    let rejections = 0;
    let abstentions = 0;
    for (const v of this.#votes) {
      if (v.choice === 'approve') approvals++;
      else if (v.choice === 'reject') rejections++;
      else abstentions++;
    }
    return {
      met: approvals >= threshold,
      threshold,
      totalVotes: this.#votes.length,
      approvals,
      rejections,
      abstentions,
    };
  }
}

/** Builder for a {@link Decision}. */
export class DecisionBuilder {
  #title: string;
  #description: string;
  #proposer: Did;
  #class?: string;

  constructor(args: {
    title: string;
    description: string;
    proposer: Did | string;
  }) {
    this.#title = args.title;
    this.#description = args.description;
    this.#proposer =
      typeof args.proposer === 'string' ? validateDid(args.proposer) : args.proposer;
  }

  /** Attach an optional decision class (free-form label). */
  public decisionClass(name: string): this {
    this.#class = name;
    return this;
  }

  /** Validate and build the {@link Decision}. */
  public async build(): Promise<Decision> {
    if (this.#title.length === 0) {
      throw new GovernanceError('title must be non-empty');
    }
    const decisionId = await computeDecisionId(
      this.#title,
      this.#description,
      this.#proposer,
    );
    const init: {
      decisionId: Hash256;
      title: string;
      description: string;
      proposer: Did;
      class?: string;
    } = {
      decisionId,
      title: this.#title,
      description: this.#description,
      proposer: this.#proposer,
    };
    if (this.#class !== undefined) {
      init.class = this.#class;
    }
    return new Decision(init);
  }
}

async function computeDecisionId(
  title: string,
  description: string,
  proposer: Did,
): Promise<Hash256> {
  const enc = new TextEncoder();
  const a = enc.encode(title);
  const b = enc.encode(description);
  const c = enc.encode(proposer);
  const payload = new Uint8Array(a.length + 1 + b.length + 1 + c.length);
  let off = 0;
  payload.set(a, off);
  off += a.length;
  payload[off++] = 0;
  payload.set(b, off);
  off += b.length;
  payload[off++] = 0;
  payload.set(c, off);
  const digest = await sha256(payload);
  return bytesToHex(digest) as Hash256;
}
